File: BOJ/[1107] 리모컨/JFe/1107.py
import sys

N = int(sys.stdin.readline())
M = int(sys.stdin.readline())
button = [int(x) for x in sys.stdin.readline().split()]
result = abs(N - 100)
# 이동 가능한 번호 탐색
for i in range(1000000):
    temp = list(str(i))
    flag = True
    # 고장난 번호가 포함되어 있으면 flag를 False로 바꾸고 break
    for t in temp:
        if int(t) in button:
            flag = False
            break
    # flag가 True면 이동하는데 누르는 횟수(len(temp)) + 이동한 위치에서 목표 위치까지 차이(abs(N - i))와 result 중 최솟값 업데이트
    if flag: result = min(result, abs(N - i) + len(temp))
print(result)

File: BOJ/[1107] 리모컨/woong-jae/1107.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const input = [];
rl.on('line', function (line) {
    input.push(...line.split(" "));
}).on("close", function () {
    solution(input[0], input[1], input.slice(2));
    process.exit();
});

function solution(N, M, broken_btns) {
    const isBrokenBtn = (() => {
        const mask = broken_btns.reduce((acc, btn) => acc | (1 << +btn), 0);
        return (btn) => (mask & (1 << btn)) > 0
    })();

    const calcPress = channel => {
        return String(channel).length + Math.abs(+N - channel);
    }

    const getMinPress = channel => {
        if (channel > 1000000) return Infinity;

        let result = calcPress(channel);
        for (let btn = channel === 0 ? 1 : 0; btn < 10; btn++) {
            if (isBrokenBtn(btn)) continue;
            result = Math.min(result, getMinPress(channel * 10 + btn));
        }
        return result;
    }

    let result = Math.abs(+N - 100);
    for (let i = 0; i < 10; i++) {
        if (isBrokenBtn(i)) continue;
        result = Math.min(result, getMinPress(i));
    }

    console.log(result);
}